Use child loggers broadly in PolicyManager

prdoyle opened the pull request:
We'd like to be able to adjust logging independently for different components / modules.

prdoyle committed:
Use child loggers broadly in PolicyManager

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

prdoyle commented:
D'oh indeed, Homer, d'oh indeed.

```
expected:<ModuleEntitlements[componentName=plugin1, entitlementsByType={}, fileAccess=org.elasticsearch.entitlement.runtime.policy.FileAccessTree@7b5089eb, logger=org.elasticsearch.common.logging.internal.LoggerImpl@33cd69a]>
but was:<ModuleEntitlements[componentName=plugin1, entitlementsByType={}, fileAccess=org.elasticsearch.entitlement.runtime.policy.FileAccessTree@7b5089eb, logger=org.elasticsearch.common.logging.internal.LoggerImpl@199d0076]>
```

rjernst commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyManager.java:
```diff
@@ -553,8 +565,7 @@ private void notEntitled(String message, Class<?> callerClass, String componentN
             var moduleName = callerClass.getModule().getName();
-            var loggerSuffix = "." + componentName + "." + ((moduleName == null) ? ALL_UNNAMED : moduleName);
-            var notEntitledLogger = LogManager.getLogger(PolicyManager.class.getName() + loggerSuffix);
+            var notEntitledLogger = getLogger(componentName, moduleName);
```
Shouldn't we already have the logger? The caller to  notEntitled should have grabbed ModuleEntitlements?

prdoyle replied:
That seems cleaner. Let me try it out.

prdoyle replied:
Yeah that's much better.

rjernst reviewed (approved):
Looks fine. One nit that could be a followup as it's not performance critical (only happens in case entitlement is being denied) and will likely require some refactoring.

prdoyle committed:
Pass ModuleEntitlements to notEntitled

prdoyle committed:
Store logger name instead of object.

Some of our unit tests check for equality of `ModuleEntitlements` objects,
and they are entitled to do so (no pun intended).

prdoyle committed:
Alright, let's cache 'em

prdoyle committed:
Memoize at point of creation, not point of use

prdoyle committed:
Explanatory comments

prdoyle committed:
Merge branch 'main' into fine-grained-loggers

prdoyle committed:
Merge branch 'main' into fine-grained-loggers

prdoyle committed:
Merge branch 'main' into fine-grained-loggers

prdoyle committed:
Merge branch 'main' into fine-grained-loggers

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.18](https://github.com/elastic/elasticsearch/pull/124937)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124937">](https://github.com/elastic/elasticsearch/pull/124937) |
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/124938)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124938">](https://github.com/elastic/elasticsearch/pull/124938) |
| ✅ |  [9.0](https://github.com/elastic/elasticsearch/pull/124939)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124939">](https://github.com/elastic/elasticsearch/pull/124939) |
